### ISSUE TITLE
Transformer-XL recurrence different from how it is presented in the paper

### ISSUE THREAD

**Issue by Atze00** [lucidrains/x-transformers#64]
The current Transformer-XL implementation uses attention length equal to the input segment length plus the memory length, while in the paper the attention length is presented as independent from the input length or the memory length. This behavior is unwanted since you can't benefit from the extended receptive field presented in figure 2. [https://arxiv.org/pdf/1901.02860.pdf](https://arxiv.org/pdf/1901.02860.pdf)
A solution could be to use an attention mask providing a further parameter to the model that automatically generates the attention mask. A snippet of code of how it could be implemented: 
```python
if self.causal:
    i, j = dots.shape[-2:]
    r = torch.arange(i, device = device)
    distance = rearrange(r, 'j -> () () () j') - rearrange(r, 'i -> () () i ()')
    mask = distance > 0
    if self.att_len:
        mask_2 = distance < self.att_len
        mask = torch.logical_and(mask, mask_2)
        del mask_2
    mask = F.pad(mask, (j - i, 0), value = False)
    dots.masked_fill_(mask, mask_value)
    del mask
```

**Comment by lucidrains** [comment]
@Atze00 i see, so you'd basically like to be able to train with different segment lengths on training vs eval - and your solution would be to cap how far each token can look into the past? (including memories)

**Comment by lucidrains** [comment]
@Atze00 using something like `max_attn_length` ?

**Comment by Atze00** [comment]
Yes, that's correct. I'd like to be able to train with different segment lengths on training vs eval, while maintaining constant how far each token can look into the past. 
The solution could be something like `max_attn_length`, which I have called `self.att_len` in the snipped of code of the first message.

**Comment by lucidrains** [comment]
@Atze00 oh gotcha, ok, i'll get this built and released this weekend!

**Comment by Atze00** [comment]
Thanks! Nice work!

**Comment by lucidrains** [comment]
@Atze00 how does this look? https://github.com/lucidrains/x-transformers/commit/55ca5d96c8b850b064177091f7a1dcfe784b24ce

**Comment by Atze00** [comment]
Looking great! Looking back, there were some problems in my snippet, thanks and good work!
